### PR TITLE
M2: ASIM Parsers, connector optimizations/enhancements

### PR DESCRIPTION
**Pull request by ashwinvenkatesha**
Required items, please complete
   
   Change(s):
   - Added ASIM parser for audit and network session schema
   -  Made connector optimizations and added an option to ingest traffic data based on policy decision

   Reason for Change(s):
   - Adding new functionality
   - Improving scope of solution

   Version Updated:
   - Required only for Detections/Analytic Rule templates

   Testing Completed:
   - See guidance below

   Checked that the validations are passing and have addressed any issues that are present:
   - See guidance below

### CODE REVIEW

**Comment by vakohl** [comment]
@ashwinvenkatesha please pull latest repo changes from upstream and check out errors listed in script job for below 4 tests.
image

e.g.

image

**Comment by ashwinvenkatesha** [comment]
This is parser with filter params for audit events
image

This is parser with filter params for network events
image

This is a different set of parsers with filter params for network events
image

Manually verified.

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventIllumioSaaSCore.yaml [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+                isnotnull(created_by.user), created_by.user.username,
+                isnotnull(created_by.agent), created_by.agent.hostname,
+                "Unknown"
+            )
+        | extend
+             ActorUsernameType = case(
+              isnotnull(created_by.system), "System",
+              isnotnull(created_by.user), "User",
+              isnotnull(created_by.agent), "Agent",
+              "Unknown"
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/ASimAuditEventIllumioSaaSCore.yaml:345]: image

Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventIllumioSaaSCore.yaml:345]: please use allowed values only:
https://learn.microsoft.com/en-us/azure/sentinel/normalization-about-schemas#the-user-name

**Review by vakohl — CHANGES_REQUESTED** [review]

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventIllumioSaaSCore.yaml [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+  parser (
+      starttime = starttime,
+      endtime = endtime,
+      srcipaddr_has_any_prefix = srcipaddr_has_any_prefix,
+      actorusername_has_any = actorusername_has_any,
+      eventtype_in = eventtype_in,
+      eventresult = eventresult,
+      operation_has_any = operation_has_any,
+      object_has_any=object_has_any,
+      newvalue_has_any=newvalue_has_any,
```
[the remark this answers is not in the document]
Comment (ashwinvenkatesha) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventIllumioSaaSCore.yaml:449]: would it be better to enumerate all the possible new value keys and based on a case statement, extract the new value?
I wanted to avoid the above approach and store the whole json object to preserve context of the entire object.
If you notice the first record, there are multiple keys, log_traffic, visibility_level, enforcement_mode and so on.
Each have a before/after key, what would be your approach to extract them?

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionIllumioSaaSCore.yaml [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+          pd,
+          ClassDetail
+  };  
+  parser(starttime=starttime, 
+  endtime=endtime, 
+  srcipaddr_has_any_prefix=srcipaddr_has_any_prefix, 
+  dstipaddr_has_any_prefix=dstipaddr_has_any_prefix, 
+  ipaddr_has_any_prefix=ipaddr_has_any_prefix,
+  dstportnumber=dstportnumber,
+  hostname_has_any=hostname_has_any,
```
[the remark this answers is not in the document]
Comment (ashwinvenkatesha) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionIllumioSaaSCore.yaml:372]: In my latest changes, I see this
image
when I run parser with 
 hostname_has_any=dynamic(['self-serve-mnc-1-vm02']), 
 
 
 change imo is
 `and (array_length(hostname_has_any)==0 or dst_hostname has_any (hostname_has_any) or src_hostname has_any(hostname_has_any))`
 
 I didnt have a check for `src_hostname has_any()` before

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventIllumioSaaSCore.yaml [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+          created_by, // used by ActorUsername and ActorType
+          pce_fqdn, // used by Dvc
+          href, // used by EventOriginalUid
+          TenantId
+  };
+  parser (
+      starttime = starttime,
+      endtime = endtime,
+      srcipaddr_has_any_prefix = srcipaddr_has_any_prefix,
+      actorusername_has_any = actorusername_has_any,
```
[the remark this answers is not in the document]
Comment (ashwinvenkatesha) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventIllumioSaaSCore.yaml:444]: > System name coming in ActorUsername. Is this expected? image

The reason System shows up is, its an event that our product has generated in response to something like, an api call made by an agent failed to authenticate.
It is expected.

[the remark this answers is not in the document]
Comment (ashwinvenkatesha) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventIllumioSaaSCore.yaml:444]: > I think you have merged Authentication logs with Audit. You need to separate out Audit and Authentication logs. Authentication schema parser should be created for Login, Logout related events. image

Will separate it out using the below condition in asim and vim parsers
```
 | where not(disabled) and (event_type !startswith "user") // filter out user auth events
```

**Comment by ashwinvenkatesha** [comment]
@vakohl 
This seems like a build pipeline issue, I have ensured to upload `*_ingestedlogs.csv`
image

However, the test fails saying
```
Starting ingestion for sample data present in Parsers/ASimAuditEvent/Parsers/ASimAuditEventIllumioSaaSCore.yaml
Error: An error occurred while trying to get content of Sample Data file located at https://raw.githubusercontent.com/Azure/Azure-Sentinel/fbbd3208270b06c07e867674eb0[30](https://github.com/Azure/Azure-Sentinel/actions/runs/11184278643/job/31094877341?pr=10994#step:8:31)9908c2cf44e/Sample%20Data/ASIM/Illumio_Core_ASimAuditEvent_IngestedLogs.csv: 404: Not Found
Ignoring this Parsers/ASimAuditEvent/Parsers/imAuditEvent.yaml because it is a union parser file
Starting ingestion for sample data present in Parsers/ASimAuditEvent/Parsers/vimAuditEventIllumioSaaSCore.yaml
Error: An error occurred while trying to get content of Sample Data file located at https://raw.githubusercontent.com/Azure/Azure-Sentinel/fbbd3208270b06c07e867674eb0309908c2cf44e/Sample%20Data/ASIM/Illumio_Core_ASimAuditEvent_IngestedLogs.csv: 404: Not Found
```

For this reason, as a workaround I had added both files
Illumio_Core_ASimAuditEvent_IngestedLogs.csv and Illumio_Core_AuditEvent_IngestedLogs.
For now, il add both types of files to let the tests pass but this the duplicate files need to be resolved.
Please advise.

File: Parsers/ASimAuthentication/Parsers/ASimAuthenticationIllumioSaaSCore.yaml [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+              EventStartTime=TimeGenerated
+              ,
+              EventEndTime=TimeGenerated
+              ,                            
+              ActorUsername = case(                  
+                  isnotnull(created_by.user), created_by.user.username,                  
+                  "Unknown"
+              ),
+              ActorUsernameType = "Simple",
+              // EventUid = _ItemId // commenting this since its failing in build pipeline
```
Comment (vakohl) [inline Parsers/ASimAuthentication/Parsers/ASimAuthenticationIllumioSaaSCore.yaml:65]: @ashwinvenkatesha 
it should not fail. If you get error related to _ItemId. Add it to your custom table

File: Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionIllumioSaaSCore.yaml [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        DstUsername = iif(dir=='I', un, '')
+      | extend
+        SrcUsernameType = _ASIM_GetUsernameType(SrcUsername),
+        DstUsernameType = _ASIM_GetUsernameType(DstUsername)
+      //Aliases
+      | extend 
+          DvcIpAddr = SrcIpAddr,
+          DvcHostname = SrcHostname
+      | extend
+          AdditionalFields = bag_pack("Class", ClassDetail,
```
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionIllumioSaaSCore.yaml:256]: @ashwinvenkatesha this is still commented in parsers

File: Parsers/ASimNetworkSession/Parsers/vimNetworkSessionIllumioSaaSCore.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionIllumioSaaSCore.yaml:382]: please use this in parser
and (array_length(hostname_has_any)==0 or dst_hostname has_any (hostname_has_any) or src_hostname has_any(hostname_has_any))

[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimNetworkSession/Parsers/vimNetworkSessionIllumioSaaSCore.yaml:382]: hostname_has_any filter was still not working

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventIllumioSaaSCore.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventIllumioSaaSCore.yaml:432]: sounds good.

File: Parsers/ASimAuthentication/Tests/Illumio_Core_Authentication_SchemaTest.csv [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+"(2) Info: Missing optional field [TargetOriginalUserType]"
+"(2) Info: Missing optional field [TargetPortNumber]"
+"(2) Info: Missing optional field [TargetRiskLevel]"
+"(2) Info: Missing optional field [TargetSessionId]"
+"(2) Info: Missing optional field [TargetUrl]"
+"(2) Info: Missing optional field [TargetUserId]"
+"(2) Info: Missing optional field [TargetUserScopeId]"
+"(2) Info: Missing optional field [TargetUserScope]"
+"(2) Info: Missing optional field [TargetUserType]"
+"(2) Info: Missing optional field [TargetUsername]"
```
Comment (vakohl) [inline Parsers/ASimAuthentication/Tests/Illumio_Core_Authentication_SchemaTest.csv:94]: image

File: Parsers/ASimAuthentication/Parsers/ASimAuthenticationIllumioSaaSCore.yaml [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            EventSchemaVersion='0.1.3'
+            ,    
+            EventOriginalUid = href
+      | lookup EventTypeLookup on event_type //fetch EventType, EventResultDetails, EventResult
+      | extend              
+              EventStartTime=TimeGenerated
+              ,
+              EventEndTime=TimeGenerated
+              ,                            
+              ActorUsername = case(
```
Comment (vakohl) [inline Parsers/ASimAuthentication/Parsers/ASimAuthenticationIllumioSaaSCore.yaml:60]: Please change this to TargetUsername
In Authentication schema,  user that got authenticated, should be the target user

File: Parsers/ASimAuthentication/Parsers/vimAuthenticationIllumioSaaSCore.yaml [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        disabled: bool=false
+    ) {
+      Illumio_Auditable_Events_CL
+      | where not(disabled) and  event_type startswith "user" and event_type !endswith "terminated" // limited to user signin, login, logoff, signoff events only
+      | where
+          (isnull(starttime) or TimeGenerated >= starttime) 
+          and (isnull(endtime) or TimeGenerated <= endtime) 
+          and (array_length(targetappname_has_any) == 0) // TargetAppName not available in source
+          and (array_length(srchostname_has_any) == 0)  // srchostname_has_any not available in source          
+          and (array_length(eventresultdetails_in) == 0) // EventResultDetails not available in source
```
Comment (vakohl) [inline Parsers/ASimAuthentication/Parsers/vimAuthenticationIllumioSaaSCore.yaml:84]: You are setting eventresultdetails_in. Please update the filter condition post doing lookup 'EventTypeLookup' 
| where (array_length(eventresultdetails_in) == 0 or EventResultDetails in~ (eventresultdetails_in))

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEvent.yaml [hunk 10]
```diff
@@ -33,6 +33,7 @@ Parsers:
   - _ASim_AuditEvent_SentinelOne
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/ASimAuditEvent.yaml:33]: Can you please include illumio details in EventVendor and EventProduct fields for the schemas you wrote parser for i.e. Audit, Network and Authentication?
https://github.com/Azure/Azure-Sentinel/blob/master/ASIM/dev/ASimTester/ASimTester.csv

Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEvent.yaml:33]: image

Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEvent.yaml:33]: image

Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEvent.yaml:33]: image

File: Sample Data/ASIM/Illumio_Core_ASimAuthentication_IngestedLogs.csv [hunk 11]
```diff
@@ -0,0 +1,73 @@
+﻿"TimeGenerated [Local Time]",Type,"_ItemId","_ResourceId",EventProduct,EventVendor,EventSchema,EventCount,EventSchemaVersion,EventType,EventResultDetails,EventResult,"EventStartTime [Local Time]","EventEndTime [Local Time]",ActorUsername,ActorUsernameType,EventUid,SrcIpAddr,Dvc,IpAddr
```
Comment (vakohl) [inline Sample Data/ASIM/Illumio_Core_ASimAuthentication_IngestedLogs.csv:1]: please remove this file

File: Sample Data/ASIM/Illumio_Core_ASimNetworkSession_IngestedLogs.csv [hunk 12]
```diff
@@ -0,0 +1,101 @@
+﻿"TimeGenerated [Local Time]",Type,"_ItemId","_ResourceId",ASimMatchingIpAddr,ASimMatchingHostname,NetworkProtocol,NetworkProtocolVersion,DvcAction,EventResult,EventCount,"EventStartTime [Local Time]","EventEndTime [Local Time]",EventType,EventProduct,EventVendor,EventSchemaVersion,EventSchema,Dvc,NetworkDirection,NetworkDuration,DstBytes,SrcBytes,DstIpAddr,SrcIpAddr,DstPortNumber,DstHostname,SrcHostname,EventSeverity,SrcProcessName,DstProcessName,SrcUsername,DstUsername,SrcUsernameType,DstUsernameType,DvcIpAddr,DvcHostname,AdditionalFields,Duration,User,Hostname,IpAddr,EventUid
```
Comment (vakohl) [inline Sample Data/ASIM/Illumio_Core_ASimNetworkSession_IngestedLogs.csv:1]: please remove this file

File: Parsers/ASimAuthentication/Tests/Illumio_Core_Authentication_SchemaTest.csv (hunk 7 above)
Comment (ashwinvenkatesha) [reply Parsers/ASimAuthentication/Tests/Illumio_Core_Authentication_SchemaTest.csv:94]: Here, TargetUserName and TargetUserNameType are the relevant fields I am including with an alias to User as well.

File: Sample Data/ASIM/Illumio_Core_AuditEvent_IngestedLogs.csv [hunk 13]
```diff
@@ -0,0 +1,122 @@
+﻿"TimeGenerated [Local Time]",href,"pce_fqdn","created_by","event_type",status,severity,action,"resource_changes",notifications,version,TenantId,Type,"_ResourceId"
```
Comment (var9287) [inline Sample Data/ASIM/Illumio_Core_AuditEvent_IngestedLogs.csv:1]: @ashwinvenkatesha there is some issue with this csv file. please see the error message.
image

File: Sample Data/ASIM/Illumio_Core_NetworkSession_IngestedLogs.csv [hunk 14]
```diff
@@ -0,0 +1,76 @@
+﻿"TimeGenerated [Local Time]","icmp_type",code,"dst_dbi","dst_dbo","dst_tbi","dst_tbo",ddms,tdms,pn,un,"src_ip","dst_ip",class,proto,"dst_port","flow_count",dir,"org_id",state,"pd_qualifier",pd,"src_hostname","src_href","dst_hostname","dst_href",network,"src_labels","dst_labels","interval_sec","pce_fqdn",version,TenantId,Type,"_ResourceId"
```
Comment (vakohl) [inline Sample Data/ASIM/Illumio_Core_NetworkSession_IngestedLogs.csv:1]: @ashwinvenkatesha issue with this csv as well.

image

Comment (vakohl) [reply Sample Data/ASIM/Illumio_Core_NetworkSession_IngestedLogs.csv:1]: please fix the csv file. You can view this error if you open the csv file in different tab
https://github.com/Azure/Azure-Sentinel/blob/fc71e73082c84a439fc5b3dd5421903810756de7/Sample%20Data/ASIM/Illumio_Core_NetworkSession_IngestedLogs.csv

File: Sample Data/ASIM/Illumio_Core_AuditEvent_IngestedLogs.csv (hunk 13 above)
Comment (vakohl) [reply Sample Data/ASIM/Illumio_Core_AuditEvent_IngestedLogs.csv:1]: https://github.com/Azure/Azure-Sentinel/blob/fc71e73082c84a439fc5b3dd5421903810756de7/Sample%20Data/ASIM/Illumio_Core_AuditEvent_IngestedLogs.csv

**Comment by vakohl** [comment]
hi @v-atulyadav @v-prasadboke ,
I've reviewed the ASim parser changes. All changes except once csv related comment is left, once that is resolved, parser changes are approved from my side. Can you please help reviewing the solution specific files in this PR?
@ashwinvenkatesha FYI

File: Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionIllumioSaaSCore.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimNetworkSession/Parsers/ASimNetworkSessionIllumioSaaSCore.yaml:256]: @ashwinvenkatesha please resolved if done on all files.

**Comment by v-prasadboke** [comment]
As discussed in the call 
Ashwin can you please move this PR to draft state

**Comment by ashwinvenkatesha** [comment]
> As discussed in the call Ashwin can you please move this PR to draft state

I have made my changes and its ready to be merged.
Please proceed with updating the azuredeploy with master branch link once merged.

**Review by v-atulyadav — APPROVED** [review]
